src/Microsoft.ML.Api/PredictionEngine.cs

```diff
@@ -150,24 +138,64 @@ public sealed class PredictionEngine<TSrc, TDst>
-                inputSchemaDefinition, outputSchemaDefinition);
+        }
+
+        private static Func<ISchema, IRowToRowMapper> StreamChecker(IHostEnvironment env, Stream modelStream)
+        {
+            env.CheckValue(modelStream, nameof(modelStream));
+            return schema =>
+            {
+                var pipe = DataViewConstructionUtils.LoadPipeWithPredictor(env, modelStream, new EmptyDataView(env, schema));
+                var transformer = new TransformWrapper(env, pipe);
+                env.CheckParam(transformer.IsRowToRowMapper, nameof(transformer), "Must be a row to row mapper");
+                return transformer.GetRowToRowMapper(schema);
+            };
         }
 
         internal PredictionEngine(IHostEnvironment env, IDataView dataPipe, bool ignoreMissingColumns,
             SchemaDefinition inputSchemaDefinition = null, SchemaDefinition outputSchemaDefinition = null)
+            : this(env, new TransformWrapper(env, env.CheckRef(dataPipe, nameof(dataPipe))), ignoreMissingColumns, inputSchemaDefinition, outputSchemaDefinition)
+        {
+        }
+
+        internal PredictionEngine(IHostEnvironment env, ITransformer transformer, bool ignoreMissingColumns,
+            SchemaDefinition inputSchemaDefinition = null, SchemaDefinition outputSchemaDefinition = null)
+            : this(env, TransformerChecker(env, transformer), ignoreMissingColumns, inputSchemaDefinition, outputSchemaDefinition)
+        {
+        }
+
+        private static Func<ISchema, IRowToRowMapper> TransformerChecker(IExceptionContext ectx, ITransformer transformer)
+        {
+            ectx.CheckValue(transformer, nameof(transformer));
+            ectx.CheckParam(transformer.IsRowToRowMapper, nameof(transformer), "Must be a row to row mapper");
+            return transformer.GetRowToRowMapper;
+        }
+
+        private PredictionEngine(IHostEnvironment env, Func<ISchema, IRowToRowMapper> makeMapper, bool ignoreMissingColumns,
+                 SchemaDefinition inputSchemaDefinition, SchemaDefinition outputSchemaDefinition)
         {
             Contracts.CheckValue(env, nameof(env));
-            var singleThreadedEnv = env.Register("SingleThreaded", conc: 1);
-            _engine = new BatchPredictionEngine<TSrc, TDst>(singleThreadedEnv, dataPipe, ignoreMissingColumns,
-                inputSchemaDefinition, outputSchemaDefinition);
+            env.AssertValue(makeMapper);
+
+            _inputRow = DataViewConstructionUtils.CreateInputRow<TSrc>(env, inputSchemaDefinition);
+            var mapper = makeMapper(_inputRow.Schema);
+            var cursorable = TypedCursorable<TDst>.Create(env, new EmptyDataView(env, mapper.Schema), ignoreMissingColumns, outputSchemaDefinition);
+            var outputRow = mapper.GetRow(_inputRow, col => true, out _disposer);
+            _outputRow = cursorable.GetRow(outputRow);
+        }
+
+        ~PredictionEngine()
+        {
+            _disposer?.Invoke();
```

Comment:
>_disposer [](start = 12, length = 9)

Just so that we don't forget. If we were planning to make `IRow` disposable, does it call for making `PredictionEngine` also disposable?

Reply:
Yeah, you're probably right about that.

I think we need to discuss this more thoroughly.

---
In reply to: [219577322](https://github.com/dotnet/machinelearning/pull/973#discussion_r219577322) [](ancestors = 219577322)